apps/opentelemetry_experimental/src/otel_otlp_logs.erl

```diff
@@ -31,41 +31,53 @@
+            empty;
+        InstrumentationScopeLogs ->
+            Attributes = otel_resource:attributes(Resource),
+            ResourceLogs = #{resource => #{attributes => otel_otlp_common:to_attributes(Attributes),
+                                           dropped_attributes_count => otel_attributes:dropped(Attributes)},
+                             scope_logs => InstrumentationScopeLogs},
+            case otel_resource:schema_url(Resource) of
+                undefined ->
+                    #{resource_logs => [ResourceLogs]};
+                SchemaUrl ->
+                    #{resource_logs => [ResourceLogs#{schema_url => SchemaUrl}]}
+            end
     end.
 
+to_proto_by_instrumentation_scope(Tab, LogHandlerConfig) ->
+    %% Even though during the export log eventss are being inserted to another table,
+    %% some late writes to the table being exported are still possible.
+    %% Thus, we can't lookup all the log events from the table and then let otel_log_handler
+    %% delete the table completely and re-create it as it will imply the risk of losing those (possible) late writes.
+    %% So, we fix the table and traverse it with ets:take/2. After that, the late writes won't be exported,
+    %% but they will be kept in the table and ready to be exported in the next exporter runs.
+    true = ets:safe_fixtable(Tab, true),
+    try
+        Key = ets:first(Tab),
+        to_proto_by_instrumentation_scope(Tab, Key, LogHandlerConfig)
+    after
+        _ = ets:safe_fixtable(Tab, false)
+    end.
 
-to_proto_by_instrumentation_scope(Logs, Config) ->
-    ScopeLogs = logs_by_scope(Logs, Config),
-    maps:fold(fun(Scope, LogRecords, Acc) ->
-                      [#{scope => otel_otlp_common:to_instrumentation_scope_proto(Scope),
-                         log_records => LogRecords
-                         %% schema_url              => unicode:chardata() % = 3, optional
-                        } | Acc]
-              end, [], ScopeLogs).
-
-
-logs_by_scope(ScopeLogs, Config) ->
-    maps:fold(fun(InstrumentationScope, Logs, Acc) ->
-                      LogRecords = [log_record(Log, Config) || Log <- Logs],
-                      Acc#{InstrumentationScope => LogRecords}
-              end, #{}, ScopeLogs).
-
+to_proto_by_instrumentation_scope(_Tab, '$end_of_table', _LogHandlerConfig) ->
+    [];
+to_proto_by_instrumentation_scope(Tab, InstrumentationScope, LogHandlerConfig) ->
+    InstrumentationScopeLogs = lists:foldl(fun({_Scope, LogEvent}, Acc) ->
+                                                   [log_record(LogEvent, LogHandlerConfig) | Acc]
+                                           end, [], ets:take(Tab, InstrumentationScope)),
+    InstrumentationScopeSpansProto = otel_otlp_common:to_instrumentation_scope_proto(InstrumentationScope),
+    [InstrumentationScopeSpansProto#{log_records => InstrumentationScopeLogs}
+    | to_proto_by_instrumentation_scope(Tab, ets:next(Tab, InstrumentationScope), LogHandlerConfig)].
```

Comment:
This is fixed with atomics counter, table size can't grow significantly higher the limit. Probably it can  have some extra residual records from the previous run in rare race conditions, when exporting and traversing finished faster than some inserts.